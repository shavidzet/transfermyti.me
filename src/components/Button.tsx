import Color from "color";
import { keyframes } from "@emotion/core";
import React from "react";
import { styled, useTheme } from "./emotion";
import Flex from "./Flex";

const spinAnimation = keyframes({
  "0%": {
    transform: "rotate(0deg)",
  },

  "100%": {
    transform: "rotate(360deg)",
  },
});

const Base = styled.button(
  {
    borderRadius: "0.375rem",
    fontSize: "1.25rem",
    fontWeight: 400,
    padding: "0.5rem 0.75rem",
    textAlign: "center",
    transition: "0.3s all linear",
    verticalAlign: "middle",
  },
  ({ theme }) => ({
    boxShadow: theme.elevation.dp1,
  }),
);

const Loading = styled.div<{ bgColor: string; fgColor: string }>(
  {
    animation: `${spinAnimation} 1.4s infinite linear`,
    borderRadius: "50%",
    fontSize: "0.75rem",
    height: "1.5rem",
    position: "relative",
    transform: "translateZ(0)",
    width: "1.5rem",

    "&:before": {
      borderRadius: "100% 0 0 0",
      content: "''",
      height: "50%",
      width: "50%",
      position: "absolute",
      top: 0,
      left: 0,
    },

    "&:after": {
      borderRadius: "50%",
      content: "''",
      height: "75%",
      width: "75%",
      margin: "auto",
      position: "absolute",
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
    },
  },
  ({ bgColor, fgColor }) => ({
    background: `linear-gradient(to right, ${fgColor} 10%, ${bgColor} 40%)`,

    "&:before": {
      background: fgColor,
    },

    "&:after": {
      background: bgColor,
    },
  }),
);

type Variant = "default" | "eggplant" | "outline" | "primary" | "secondary";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  disabled?: boolean;
  loading?: boolean;
  variant?: Variant;
  onClick: VoidFunction;
}

const Button: React.FC<ButtonProps> = ({
  children,
  disabled = false,
  loading = false,
  variant = "default",
  onClick,
  ...props
}) => {
  const { colors } = useTheme();
  const mainColorByVariant = {
    default: colors.midnight,
    eggplant: colors.eggplant,
    outline: colors.secondary,
    primary: colors.primary,
    secondary: colors.secondary,
  }[variant];

  const altColorByVariant = {
    default: colors.secondary,
    eggplant: colors.secondary,
    outline: colors.primary,
    primary: colors.secondary,
    secondary: colors.primary,
  }[variant];

  const mainColor = Color(mainColorByVariant);
  const altColor = Color(altColorByVariant);

  const isOutline = variant === "outline";
  // Don't change the background color on hover if these conditions are met:
  const isHoverInitial = isOutline || disabled || variant === "secondary";

  return (
    <Base
      aria-disabled={disabled}
      css={{
        background: mainColor.hex(),
        border: isOutline ? `3px solid ${altColor.hex()}` : "none",
        color: altColor.hex(),
        opacity: disabled ? 0.5 : 1,

        "&:hover": {
          background: isHoverInitial
            ? mainColor.hex()
            : mainColor.darken(0.2).hex(),
          textDecoration: disabled ? "none" : "underline",
        },

        "&:focus": {
          outlineColor:
            variant === "secondary" ? colors.secondary : colors.primary,
        },
      }}
      disabled={disabled}
      onClick={disabled || loading ? undefined : onClick}
      type="button"
      {...props}
    >
      {loading ? (
        <Flex
          alignItems="center"
          justifyContent="center"
          data-testid="button-loading"
        >
          <Loading bgColor={mainColor.hex()} fgColor={altColor.hex()} />
        </Flex>
      ) : (
        children
      )}
    </Base>
  );
};

export default Button;
