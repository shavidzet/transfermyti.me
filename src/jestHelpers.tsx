/* eslint-disable import/no-extraneous-dependencies,@typescript-eslint/no-explicit-any,no-console */
import React from "react";
import {
  RenderOptions,
  RenderResult,
  render as testRender,
} from "@testing-library/react";
import { ThemeProvider } from "emotion-theming";
import { theme } from "./components/emotion";

// These are exported for convenience (so we don't have to add an additional import
// to the test file).
export { createEvent, fireEvent, RenderResult } from "@testing-library/react";
export { theme } from "./components/emotion";

/**
 * Renders the specified component with react-testing-library wrapped in a theme
 * provider so we don't have to do it explicitly for each component.
 */
export function render(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
): RenderResult {
  return testRender(<ThemeProvider theme={theme}>{ui}</ThemeProvider>, options);
}

type LogMethod = "log" | "warn" | "error";

export class ConsoleToggle {
  private consoleMethod: (...args: any[]) => void = console.warn;
  private readonly logMethod: LogMethod;

  public constructor(logMethod: LogMethod) {
    this.logMethod = logMethod;
  }

  public onBefore(): void {
    this.consoleMethod = console[this.logMethod];
    console[this.logMethod] = jest.fn();
  }

  public onAfter(): void {
    console[this.logMethod] = this.consoleMethod;
  }
}
