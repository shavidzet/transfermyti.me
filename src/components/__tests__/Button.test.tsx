import Color from "color";
import cases from "jest-in-case";
import React from "react";
import { fireEvent, render, RenderResult, theme } from "~/jestHelpers";
import Button from "../Button";

const setup = (
  propOverrides: any = {},
): { props: any; wrapper: RenderResult } => {
  const props = {
    disabled: false,
    loading: false,
    onClick: jest.fn(),
    ...propOverrides,
  };

  const wrapper = render(<Button {...props} />);

  return { props, wrapper };
};

describe("the <Button> component", () => {
  cases(
    "matches its snapshot based on props.variant",
    options => {
      const { wrapper } = setup({ variant: options.variant });

      expect(wrapper.getByRole("button")).toMatchSnapshot();
    },
    [
      {
        name: "when props.variant = default",
        variant: "default",
      },
      {
        name: "when props.variant = eggplant",
        variant: "eggplant",
      },
      {
        name: "when props.variant = outline",
        variant: "outline",
      },
      {
        name: "when props.variant = primary",
        variant: "primary",
      },
      {
        name: "when props.variant = secondary",
        variant: "secondary",
      },
    ],
  );

  test(`uses "default" as variant when props.variant is undefined`, () => {
    const { wrapper } = setup({ variant: undefined });
    const expected = Color(theme.colors.midnight).hex();

    expect(wrapper.getByRole("button")).toHaveStyleRule("background", expected);
  });

  test(`uses false for props.disabled and props.loading if undefined`, () => {
    const { wrapper } = setup({ disabled: undefined, loading: undefined });

    expect(wrapper.getByRole("button")).not.toBeDisabled();
    expect(wrapper.queryByTestId("button-loading")).toBeNull();
  });

  test("disables the button when props.disabled = true", () => {
    const { wrapper } = setup({ disabled: true });

    expect(wrapper.getByRole("button")).toHaveAttribute(
      "aria-disabled",
      "true",
    );
  });

  test("shows the loading indicator when props.loading = true", () => {
    const { wrapper } = setup({ loading: true });

    expect(wrapper.queryByTestId("button-loading")).not.toBeNull();
  });

  test("fires props.onClick when the button is clicked", () => {
    const { wrapper, props } = setup();
    fireEvent.click(wrapper.getByRole("button"));

    expect(props.onClick).toHaveBeenCalled();
  });
});
