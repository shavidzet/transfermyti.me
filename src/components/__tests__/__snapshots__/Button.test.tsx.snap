// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the <Button> component matches its snapshot based on props.variant when props.variant = default 1`] = `
.emotion-0 {
  border-radius: 0.375rem;
  font-size: 1.25rem;
  font-weight: 400;
  padding: 0.5rem 0.75rem;
  text-align: center;
  -webkit-transition: 0.3s all linear;
  transition: 0.3s all linear;
  vertical-align: middle;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.2), 0 1px 1px 0 rgba(0,0,0,0.14), 0 2px 1px -1px rgba(0,0,0,0.12);
  background: #484848;
  border: none;
  color: #EFFDE8;
  opacity: 1;
}

.emotion-0:hover {
  background: #3A3A3A;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:focus {
  outline-color: rgb(30,120,161);
}

<button
  aria-disabled="false"
  class="emotion-0"
  type="button"
/>
`;

exports[`the <Button> component matches its snapshot based on props.variant when props.variant = eggplant 1`] = `
.emotion-0 {
  border-radius: 0.375rem;
  font-size: 1.25rem;
  font-weight: 400;
  padding: 0.5rem 0.75rem;
  text-align: center;
  -webkit-transition: 0.3s all linear;
  transition: 0.3s all linear;
  vertical-align: middle;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.2), 0 1px 1px 0 rgba(0,0,0,0.14), 0 2px 1px -1px rgba(0,0,0,0.12);
  background: #550055;
  border: none;
  color: #EFFDE8;
  opacity: 1;
}

.emotion-0:hover {
  background: #440044;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:focus {
  outline-color: rgb(30,120,161);
}

<button
  aria-disabled="false"
  class="emotion-0"
  type="button"
/>
`;

exports[`the <Button> component matches its snapshot based on props.variant when props.variant = outline 1`] = `
.emotion-0 {
  border-radius: 0.375rem;
  font-size: 1.25rem;
  font-weight: 400;
  padding: 0.5rem 0.75rem;
  text-align: center;
  -webkit-transition: 0.3s all linear;
  transition: 0.3s all linear;
  vertical-align: middle;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.2), 0 1px 1px 0 rgba(0,0,0,0.14), 0 2px 1px -1px rgba(0,0,0,0.12);
  background: #EFFDE8;
  border: 3px solid #1E78A1;
  color: #1E78A1;
  opacity: 1;
}

.emotion-0:hover {
  background: #EFFDE8;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:focus {
  outline-color: rgb(30,120,161);
}

<button
  aria-disabled="false"
  class="emotion-0"
  type="button"
/>
`;

exports[`the <Button> component matches its snapshot based on props.variant when props.variant = primary 1`] = `
.emotion-0 {
  border-radius: 0.375rem;
  font-size: 1.25rem;
  font-weight: 400;
  padding: 0.5rem 0.75rem;
  text-align: center;
  -webkit-transition: 0.3s all linear;
  transition: 0.3s all linear;
  vertical-align: middle;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.2), 0 1px 1px 0 rgba(0,0,0,0.14), 0 2px 1px -1px rgba(0,0,0,0.12);
  background: #1E78A1;
  border: none;
  color: #EFFDE8;
  opacity: 1;
}

.emotion-0:hover {
  background: #186081;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:focus {
  outline-color: rgb(30,120,161);
}

<button
  aria-disabled="false"
  class="emotion-0"
  type="button"
/>
`;

exports[`the <Button> component matches its snapshot based on props.variant when props.variant = secondary 1`] = `
.emotion-0 {
  border-radius: 0.375rem;
  font-size: 1.25rem;
  font-weight: 400;
  padding: 0.5rem 0.75rem;
  text-align: center;
  -webkit-transition: 0.3s all linear;
  transition: 0.3s all linear;
  vertical-align: middle;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.2), 0 1px 1px 0 rgba(0,0,0,0.14), 0 2px 1px -1px rgba(0,0,0,0.12);
  background: #EFFDE8;
  border: none;
  color: #1E78A1;
  opacity: 1;
}

.emotion-0:hover {
  background: #EFFDE8;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:focus {
  outline-color: rgb(239,253,232);
}

<button
  aria-disabled="false"
  class="emotion-0"
  type="button"
/>
`;
